### ISSUE TITLE
Translating `if torch.sum(gt_score) < 1:  return torch.sum(pred_score + pred_geo) * 0`

### ISSUE THREAD

**Issue by mdrokz** [LaurentMazare/tch-rs#571]
Hi im working on converting this https://github.com/SakuraRiven/EAST to rust using tch-rs, 

But im struggling to convert this line of code using tch-rs https://github.com/SakuraRiven/EAST/blob/cec7ae98f9c21a475b935f74f4c3969f3a989bd4/loss.py#L31-L32 

I want it to return from the function based on that condition any ideas ?

**Comment by LaurentMazare** [comment]
I'm not really sure what this python snippet tries to do as the multiplication by 0 looks a bit odd.
Anyway, you can get the sum of elements in a tensor, equivalent to `torch.sum(gt_score)` using `gt_score.sum(tch::Kind::Float)`. This will return a tensor with a single element, if you want to extract a float so that you can compare its value with 0.5, `f64::from(&mytensor)` would do this.

**Comment by mdrokz** [comment]
> I'm not really sure what this python snippet tries to do as the multiplication by 0 looks a bit odd. Anyway, you can get the sum of elements in a tensor, equivalent to `torch.sum(gt_score)` using `gt_score.sum(tch::Kind::Float)`. This will return a tensor with a single element, if you want to extract a float so that you can compare its value with 0.5, `f64::from(&mytensor)` would do this.

Hello thank you for reply, i tried what you suggested and it works ! i had one more doubt about this line of code and im unsure if tch-rs has these functions https://github.com/SakuraRiven/EAST/blob/cec7ae98f9c21a475b935f74f4c3969f3a989bd4/model.py#L42-L52 is this possible to translate to tch-rs ?

**Comment by LaurentMazare** [comment]
This works slightly differently in `tch`: weight initialisation is specified when creating the modules as part of the modules configs. For example, `BatchNormConfig` one can set `ws_init` and `bs_init`, see [here](https://github.com/LaurentMazare/tch-rs/blob/66a0fe916fe5b65fcec6156d8d15ae43ece64e0e/src/nn/batch_norm.rs#L12-L13). Then you would create the batch-norm2d module via something like:
```rust
let bn_cfg = nn::BatchNormConfig {
    ws_init: nn::Init::Randn { mean: 0. ,stdev: 0.01 },
    bs_init: nn::Init::Const(0.),
    ..Default::default(),
};
let bn = nn::batch_norm2d(&p / "bn", c_out, bn_cfg);
```

**Comment by mdrokz** [comment]
> This works slightly differently in `tch`: weight initialisation is specified when creating the modules as part of the modules configs. For example, `BatchNormConfig` one can set `ws_init` and `bs_init`, see [here](https://github.com/LaurentMazare/tch-rs/blob/66a0fe916fe5b65fcec6156d8d15ae43ece64e0e/src/nn/batch_norm.rs#L12-L13). Then you would create the batch-norm2d module via something like:
> 
> ```rust
> let bn_cfg = nn::BatchNormConfig {
>     ws_init: nn::Init::Randn { mean: 0. ,stdev: 0.01 },
>     bs_init: nn::Init::Const(0.),
>     ..Default::default(),
> };
> let bn = nn::batch_norm2d(&p / "bn", c_out, bn_cfg);
> ```

Ah i see interesting, also is `tch::nn::Init::KaimingUniform` same as kaiming_normal ? and can we set `mode='fan_out', nonlinearity='relu'` these options for kaiming_normal ?

**Comment by LaurentMazare** [comment]
`nn.init.normal_` is I think equivalent to `nn::Init::Randn` (see [here](https://github.com/LaurentMazare/tch-rs/blob/main/src/nn/init.rs#L10)). Kaiming uniform is different from Kaiming normal as it uses a uniform distribution rather than a normal one, adding Kaiming normal should be pretty straightforward if that's helpful but maybe using the default initialization will work out of the box? (not sure why this is overriden here)

**Comment by mdrokz** [comment]
> `nn.init.normal_` is I think equivalent to `nn::Init::Randn` (see [here](https://github.com/LaurentMazare/tch-rs/blob/main/src/nn/init.rs#L10)). Kaiming uniform is different from Kaiming normal as it uses a uniform distribution rather than a normal one, adding Kaiming normal should be pretty straightforward if that's helpful but maybe using the default initialization will work out of the box? (not sure why this is overriden here)

I see, how would one go about implementing Kaiming normal is there any reference?

**Comment by LaurentMazare** [comment]
I've actually reworked the Kaiming initialization part in #573, this will allow for using Kaiming normal (as well as fix some small bug in Kaiming uniform).

When it comes to `state_dict`, the weight values can usually be accessed directly if needed. If the goal is to load or save the weights, it's usually simpler to go through the `VarStore` api.

**Comment by mdrokz** [comment]
> I've actually reworked the Kaiming initialization part in #573, this will allow for using Kaiming normal (as well as fix some small bug in Kaiming uniform).
> 
> When it comes to `state_dict`, the weight values can usually be accessed directly if needed. If the goal is to load or save the weights, it's usually simpler to go through the `VarStore` api.

Okay cool, i see the PR has been merged will the new release come soon ? for now i will use the master branch and is there any way to iterate over `Sequential` modules ?

**Comment by LaurentMazare** [comment]
The new releases are usually done on new PyTorch versions, e.g. 0.9.0 was released for PyTorch 1.13, so that may be a month or two away. If you can use the master branch until then (or the specific commit at which the PR landed) that's probably the easiest.

Not sure what you want to do with the modules inside a `Sequential` but it's better if you keep track of them explicitely as the `Sequential` will lose track of their types anyway.

**Comment by mdrokz** [comment]
> The new releases are usually done on new PyTorch versions, e.g. 0.9.0 was released for PyTorch 1.13, so that may be a month or two away. If you can use the master branch until then (or the specific commit at which the PR landed) that's probably the easiest.
> 
> Not sure what you want to do with the modules inside a `Sequential` but it's better if you keep track of them explicitely as the `Sequential` will lose track of their types anyway.

Ah i see, yep for now i will use the specific commit and thank you so much for your help i will be closing this now.
